soh/soh/Enhancements/controls/GameControlEditor.cpp

```diff
@@ -216,36 +215,132 @@ namespace GameControlEditor {
         ImGui::EndTable();
     }
 
+    // Copied from InputEditor::DrawButton
+    // CurrentPort is indexed started at 1 here due to the Generic tab
+    void DrawButton(const char* label, int n64Btn) {
+        auto controlDeck = Ship::Window::GetInstance()->GetControlDeck();
+        auto backend = controlDeck->GetPhysicalDeviceFromVirtualSlot(CurrentPort - 1);
+        float size = 40;
+        bool readingMode = BtnReading == n64Btn;
+        bool disabled = (BtnReading != -1 && !readingMode) || !backend->Connected() || backend->GetGuid() == "Auto";
+        ImVec2 len = ImGui::CalcTextSize(label);
+        ImVec2 pos = ImGui::GetCursorPos();
+        ImGui::SetCursorPosY(pos.y + len.y / 4);
+        ImGui::SetCursorPosX(pos.x + abs(len.x - size));
+        ImGui::Text("%s", label);
+        ImGui::SameLine();
+        ImGui::SetCursorPosY(pos.y);
+
+        if (disabled) {
+            ImGui::PushItemFlag(ImGuiItemFlags_Disabled, true);
+            ImGui::PushStyleVar(ImGuiStyleVar_Alpha, ImGui::GetStyle().Alpha * 0.5f);
+        }
+
+        if (readingMode) {
+            const int32_t btn = backend->ReadRawPress();
+
+            if (btn != -1) {
+                backend->SetButtonMapping(CurrentPort - 1, n64Btn, btn);
+                BtnReading = -1;
+
+                // avoid immediately triggering another button during gamepad nav
+                ImGui::SetKeyboardFocusHere(0);
+            }
+        }
+
+        const std::string BtnName = backend->GetButtonName(CurrentPort - 1, n64Btn);
+
+        if (ImGui::Button(StringHelper::Sprintf("%s##HBTNID_%d", readingMode ? "Press a Key..." : BtnName.c_str(), n64Btn).c_str())) {
+            BtnReading = n64Btn;
+            backend->ClearRawPress();
+        }
+
+        if (disabled) {
+            ImGui::PopItemFlag();
+            ImGui::PopStyleVar();
+        }
+    }
+
+    // Controller dropdown copied from InputEditor::DrawControllerSchema
```

Comment:
any chance we could reuse code for this instead?

Reply:
Not at the moment; see above comment